usersession/agent/rest_api.go

```diff
@@ -334,3 +362,179 @@ func postRefreshFinishedNotification(c *Command, r *http.Request) Response {
+		return resp
+	}
+
+	decoder := json.NewDecoder(r.Body)
+
+	// pendingSnapRefreshInfo holds information about pending snap refresh provided by snapd.
+	var refreshInfo client.DeferredSnapRefreshInfo
+	if err := decoder.Decode(&refreshInfo); err != nil {
+		return BadRequest("cannot decode request body into pending snap deferred refresh info: %v", err)
+	}
+
+	// this will just close the "working on it" window, so we also must show an extra message
+	// to the user notifying they that the update has concluded. This is even more important
+	// in the case where the user closed the "working on it" window.
+	if err := tryNotifyRefreshToSnapDesktopIntegration(refreshInfo.InstanceName, DestroyNotification); err != nil {
+		logger.Noticef("Failed to communicate with Snapd-Desktop-Integration to close a refresh popup: %v", err)
+	}
+	// TODO: this message needs to be crafted better as it's the only thing guaranteed to be delivered.
+	summary := fmt.Sprintf(i18n.G("Refreshed %q snap"), refreshInfo.InstanceName)
+	body := i18n.G("You can launch it again now.")
+
+	return sendDesktopStandardNotification(c, refreshInfo, summary, body)
+}
+
+func sendDesktopStandardNotification(c *Command, refreshInfo client.DeferredSnapRefreshInfo, summary string, body string) Response {
+	// Note that since the connection is shared, we are not closing it.
+	if c.s.bus == nil {
+		return SyncResponse(&resp{
+			Type:   ResponseTypeError,
+			Status: 500,
+			Result: &errorResult{
+				Message: "cannot connect to the session bus",
+			},
+		})
+	}
+
+	var icon string
+	var hints []notification.Hint
+
+	hints = append(hints, notification.WithUrgency(notification.NormalUrgency))
+	// The notification is provided by snapd session agent.
+	hints = append(hints, notification.WithDesktopEntry("io.snapcraft.SessionAgent"))
+	// But if we have a desktop file of the busy application, use that apps's icon.
+	if refreshInfo.AppDesktopEntry != "" {
+		parser := goconfigparser.New()
+		desktopFilePath := filepath.Join(dirs.SnapDesktopFilesDir, refreshInfo.AppDesktopEntry+".desktop")
+		if err := parser.ReadFile(desktopFilePath); err == nil {
+			icon, _ = parser.Get("Desktop Entry", "Icon")
+		}
+	}
+
+	msg := &notification.Message{
+		AppName: refreshInfo.AppName,
+		Title:   summary,
+		Icon:    icon,
+		Body:    body,
+		Hints:   hints,
+	}
+
+	// TODO: silently ignore error returned when the notification server does not exist.
+	// TODO: track returned notification ID and respond to actions, if supported.
+	if err := c.s.notificationMgr.SendNotification(notification.ID(refreshInfo.InstanceName), msg); err != nil {
+		return SyncResponse(&resp{
+			Type:   ResponseTypeError,
+			Status: 500,
+			Result: &errorResult{
+				Message: fmt.Sprintf("cannot send notification message: %v", err),
+			},
+		})
+	}
+	return SyncResponse(nil)
+}
+
+func tryUpdateRefreshStatusDesktopIntegration(snapName string, barText string, progress float64) bool {
```

Comment:
This is still returning a bool, is that intentional?

Reply:
Nope, my fault. Fixing...

Reply:
(It's a different function, I forgot to change it).

Reply:
Done.